climada/util/calibrate/bayesian_optimizer.py

```diff
@@ -0,0 +1,448 @@
+        This may happen if the parameter space is tightly bound or constrained. Defaults
+        to ``True``.
+    bayes_opt_kwds : dict
+        Additional keyword arguments passed to the ``BayesianOptimization`` constructor.
+
+    Notes
+    -----
+    The following requirements apply to the parameters of :py:class:`Input` when using
+    this class:
+
+    bounds
+        Setting ``bounds`` in the ``Input`` is required because the optimizer first
+        "explores" the bound parameter space and then narrows its search to regions
+        where the cost function is low.
+    constraints
+        Must be an instance of ``scipy.minimize.LinearConstraint`` or
+        ``scipy.minimize.NonlinearConstraint``. See
+        https://github.com/bayesian-optimization/BayesianOptimization/blob/master/examples/constraints.ipynb
+        for further information. Supplying contraints is optional.
+
+    Attributes
+    ----------
+    optimizer : bayes_opt.BayesianOptimization
+        The optimizer instance of this class.
+    """
+
+    verbose: InitVar[int] = 1
+    random_state: InitVar[int] = 1
+    allow_duplicate_points: InitVar[bool] = True
+    bayes_opt_kwds: InitVar[Optional[Mapping[str, Any]]] = None
+
+    def __post_init__(
+        self, verbose, random_state, allow_duplicate_points, bayes_opt_kwds
+    ):
+        """Create optimizer"""
+        if bayes_opt_kwds is None:
+            bayes_opt_kwds = {}
+
+        if self.input.bounds is None:
+            raise ValueError("Input.bounds is required for this optimizer")
+
+        self.optimizer = BayesianOptimization(
+            f=self._opt_func,
+            pbounds=self.input.bounds,
+            constraint=self.input.constraints,
+            verbose=verbose,
+            random_state=random_state,
+            allow_duplicate_points=allow_duplicate_points,
+            **bayes_opt_kwds,
+        )
+
+    def _target_func(self, true: pd.DataFrame, predicted: pd.DataFrame) -> Number:
+        """Invert the cost function because BayesianOptimization maximizes the target"""
+        return -self.input.cost_func(true, predicted)
+
+    def run(self, **opt_kwargs) -> BayesianOptimizerOutput:
+        """Execute the optimization
+
+        ``BayesianOptimization`` *maximizes* a target function. Therefore, this class
+        inverts the cost function and used that as target function. The cost function is
+        still minimized.
+
+        Parameters
+        ----------
+        init_points : int, optional
+            Number of initial samples taken from the parameter space. Defaults to 10^N,
+            where N is the number of parameters.
+        n_iter : int, optional
+            Number of iteration steps after initial sampling. Defaults to 10^N, where N
+            is the number of parameters.
+        opt_kwargs
+            Further keyword arguments passed to ``BayesianOptimization.maximize``.
```

Comment:
The `maximize` method now only accepts the three arguments `init_points`, `n_iter` and `acquisition_function`: https://github.com/bayesian-optimization/BayesianOptimization/pull/406

Therefore, we should probably drop the `opt_kwargs` here.

Reply:
Not sure. I would like to keep the option to add further arguments. However, I also do not want to suggest users have to change the acquisition function. So I would rather keep it as is, which is also relatively safe in terms of forward- and backward-compatibility.